### ISSUE TITLE
Select rest after match

### ISSUE THREAD

**Issue by hugobessaa** [nathanmarz/specter#51]
Hey :smile: 

I'm trying to use `specter` to test hiccup structures. I would like to know if there's a way to select the rest of the expression after finding a match.

Example:
```cljs
(def hiccup [:div [:a "content"]])

(select-first [(walker vector?) (filterer #(= :a (first %))) some-selector] hiccup)
;; "content"
```

This is useful to test my `reagent` components without rendering them.

**Comment by hugobessaa** [comment]
Reading a little more about specter, I found `TreeWalker` in the docs.

This may solve my problem:
```cljs
(extend-protocolpath TreeWalker
                     Object nil
                     clojure.lang.PersistentVector [ALL TreeWalker])

(sp/select-first [TreeWalker
                  (sp/filterer #(= :div (first %)))]
                 [[:a [:div "one"]]])
;; [:div "one"]
```

The problem is that the `com.rpl.specter.macros` namespace is `clj` only. Would it be difficult to port it to `cljx`?

**Comment by nathanmarz** [comment]
I'm assuming you're using ClojureScript here – ClojureScript macros are always done in Clojure. The macros in that namespace will work fine.

Doesn't look to me that code will work, as TreeWalker will navigate you to :a, :div, and "one". You can make your own walker that looks something like this:

```clojure
(extend-protocolpath MyWalker
                     Object VOID
                     clojure.lang.PersistentVector (if-path [FIRST #(= :div %)] nil [ALL MyWalker]))
```

This uses VOID which isn't released yet, but you can just copy the definition from the source until 0.9.2 is released: https://github.com/nathanmarz/specter/blob/master/src/clj/com/rpl/specter.cljx#L138

**Comment by hugobessaa** [comment]
Thank you for your fast response. Been using specter and loving it.

Yes, I'm using ClojureScript. For some reason I'm not able to load `extend-protocolpath`. I'll take a look at what is happening.

After some fiddling I found this solution:

```cljs
(defn selector-walker
  [selector]
  (fn [x]
    (and (vector? x)
         (= selector (first x)))))

(select [(walker (selector-walker :div))]
        [:.some [:nested [:elements [:div "content"]]]]) 
;; => ([:div "content"])
```

However, it doesn't work if you have nested things you want to select, just siblings. This is the expected behavior when using `walker`.

```cljs
(select [(walker (selector-walker :div))]
        [:.some [:nested [:elements [:div [:div "oh my"]]]]]) 
;; => ([:div [:div "content"]])
```

**Comment by nathanmarz** [comment]
Oh sorry I gave you bad information. Protocol paths are not available in ClojureScript (see #39). Without those, you'll have to iteratively call your selector on the results returned from the selector to continue navigating to nested divs. I've also opened #52 as another possible solution to this problem.

In terms of how to solve your problem with protocol paths, I had the idea for a selector that selects the current element as well as navigating further beyond that element. With this `continue-then-stay` selector you can solve your problem like this:

```clojure
(defn continue-then-stay
  [& path]
  (fixed-pathed-path [late path]
    (select* [this structure next-fn]
      (concat (doall (mapcat next-fn (compiled-select late structure)))
              (next-fn structure)))
    (transform* [this structure next-fn]
      (next-fn (compiled-transform late next-fn structure))
      )))

(defprotocolpath MyWalker [])

(extend-protocolpath
  MyWalker
  Object VOID

  clojure.lang.PersistentVector
  (if-path [FIRST #(= :div %)]
    (continue-then-stay ALL MyWalker)
    [ALL MyWalker]
    ))

(select MyWalker
        [:.some [:nested [:elements [:div [:div "oh my"]]]]])
;; => ([:div "oh my"] [:div [:div "oh my"]])
```

While `continue-then-stay` is effectively pre-order traversal, a similar `stay-then-continue` version would implement post-order traversal.

**Comment by nathanmarz** [comment]
With `declarepath` functionality, your selector would look something like:

```clojure
(declarepath MyWalker)

(def MyWalker
  (if-path vector?
    (if-path [FIRST #(= :div %)]
      (continue-then-stay ALL MyWalker)
      [ALL MyWalker]
      )))
```

**Comment by nathanmarz** [comment]
Turns out everything I mentioned to solve your problem was pretty trivial to implement, so I implemented it and pushed 0.9.2-SNAPSHOT with the new functionality. Here's code that solves your original problem and handles nested divs without any problems:

```clojure
(require '[com.rpl.specter :as s])
(require-macros '[com.rpl.specter.macros :refer [declarepath providepath]])

(declarepath MyWalker)

(providepath MyWalker
  (s/if-path vector?
    (s/if-path [s/FIRST #(= :div %)]
      (s/continue-then-stay s/ALL MyWalker)
      [s/ALL MyWalker]
      )))

(s/select MyWalker
  [:.some [:nested [:elements [:div [:div "oh my"]]]]])
;; => ([:div "oh my"] [:div [:div "oh my"]])
```

**Comment by hugobessaa** [comment]
Ah BIG thank you, @nathanmarz! I will be using this a lot.
